### PR TITLE
Add docs to hover.

### PR DESCRIPTION
**Pull request by ChrisPenner**
Note:

This involves rendering docs so it's affected by #3879;  After IOSource is cached everything works fine. :'( ; I added a timeout so we can still render type-signatures in the meantime, but it'll work better once that regression is patched 😄

If you want to test this out, wait about 15s after UCM is booted first.

See the transcript for an example of how it renders

## Overview

Show docs in the help when you hover on something.

https://user-images.githubusercontent.com/6439644/227071689-15a8e5e8-01df-4cd7-a61b-01734fb86877.mp4

## Implementation notes

We need a Markdown renderer for docs, which was slightly more annoying than I anticipated haha.

I originally tried using the cmark GFM renderer. Unfortunately, the renderer is needlessly strict about a lot of things, and it was like pulling teeth to get it to not just crash because things got nested in a way it didn't like, so I found it much easier to write a custom renderer that works exactly how I want.

* Adds compilation of IOSource's typechecked file to the things we do on startup so it's ready when we need it. We can maybe remove this if #3880  is resolved.

## Test coverage

* Added the `debug.doc-to-markdown` command
* And a transcript which tests rendering a doc with all the relevant doc features.

## Loose ends

- It turns out we don't actually have parsers for a lot of things we have constructors for, maybe we should add those at some point (e.g. tables, images, etc.)
- Currently type-links and doc-links are rendered as plain text. I _think_ there are ways to support these in the LSP format, but I'll look into that in a later change.

### CODE REVIEW

**Comment by hojberg** [comment]
@ChrisPenner will this do truncations of the docs in some way? Some of the core docs are really substantial—thats great, but might be overwhelming for a tooltip.

**Comment by ChrisPenner** [comment]
@hojberg I think it's generally the Client's job to truncate or scroll or whatever it needs to do 🤔  

I added a video which show's VS Code's scrolling behaviour which seems nice to me :)

File: unison-cli/src/Unison/CommandLine/Main.hs [hunk 1]
```diff
@@ -116,10 +117,12 @@ main dir welcome initialPath config initialInputs runtime sbRuntime codebase ser
...
       -- overwrite it.
       void $ tryPutTMVar rootVar root
-    -- Start forcing the thunk in a background thread.
+    -- Start forcing thunks in a background thread.
     -- This might be overly aggressive, maybe we should just evaluate the top level but avoid
     -- recursive "deep*" things.
-    void $ UnliftIO.evaluate root
+    UnliftIO.concurrently_
+      (UnliftIO.evaluate root)
+      (UnliftIO.evaluate typecheckedFile) -- IOSource takes a while to compile, we should start compiling it on startup
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/CommandLine/Main.hs]: Potentially we can remove this if it compiles on-demand fast enough, but it's also probably fine to leave here.

File: unison-cli/src/Unison/LSP.hs [hunk 2]
```diff
@@ -136,10 +137,11 @@ lspDoInitialize vfsVar codebase runtime scope latestBranch latestPath lspContext
   currentPathCacheVar <- newTVarIO Path.absoluteEmpty
   cancellationMapVar <- newTVarIO mempty
   completionsVar <- newTVarIO mempty
-  let env = Env {ppedCache = readTVarIO ppedCacheVar, parseNamesCache = readTVarIO parseNamesCacheVar, currentPathCache = readTVarIO currentPathCacheVar, ..}
+  nameSearchCacheVar <- newTVarIO $ Backend.makeNameSearch 0 mempty
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/LSP.hs:140]: Add a name search to the LSP context, updated when UCM changes namespaces.

File: unison-share-api/src/Unison/Server/Doc/Markdown/Render.hs
Comment (ChrisPenner) [inline unison-share-api/src/Unison/Server/Doc/Markdown/Render.hs:1]: I copied the structure from the existing Doc -> HTML renderer, then tweaked it a bunch to render to a markdown type, then I render that markdown type to text.

**Review by ceedubs — APPROVED** [review]
👍 I gave this a try and it worked beautifully (at least after the delay that you warned about).

**Review by pchiusano — APPROVED** [review]

**Pull request merged**